design/automatic-network-configuration.md

```diff
@@ -0,0 +1,355 @@
+    `spec.neighborRef`.
+
+#### Network CRD
+
+```yaml
+apiVersion: network.metal3.io/v1alpha1
+kind: Network
+metaData:
+  name: sample-network
+spec:
+  networkType: physicalSwitch
+  // The VLANs available
+  vlans:
+    - id:
+    - id:
+  // The untagged VLAN ID
+  vlanId:
+  interfaces:
+    - name: bm0-eth0
+    - name: bm1-eth1
+```
+
+#### Network Controller
+- Watch
+  - Network CR
+- Configuration Operations
+  - When the network configuration is changed, finds the corresponding
+    SwitchPort CR according to `.spec.neighborRef` in each Interface CR which
+    is stored in `.spec.interfaces` in the Network CR. Then changes the spec of
+    all the SwitchPort CRs.
+  - When `.spec.interfaces` increases, finds the corresponding Interface CR,
+    and then finds the corresponding SwitchPort CR, and write configuration to
+    that SwitchPort CR.
+
+#### Switch CRD
+The Switch CR is used to store the specific information of the switch including
+its name and access details.
+```yaml
+apiVersion: switch.metal3.io/v1alpha1
+kind: Switch
+metadata:
+  name: example-switch
+spec:
+  accessDetail:
+    ip: 192.0.0.1
+    OS: fos
+    port: 22
+    // Store the username and password of the switch
+    // in the form of secret
+    accessSecret: example-switch-secret
+```
+
+#### SwitchPort CRD
+SwitchPort CR contains the desired configuration and current status of a
+specific port in a switch.
+```yaml
+apiVersion: switch.metal3.io/v1alpha1
+kind: SwitchPort
+metadata:
+  name: example-switchport
+spec:
+  portId: 10
+  // vlans indicates the vlan information to be changed to.
+  // Changed by network controller.
+  // The VLANs available
+  vlans:
+    - id:
+    - id:
+  // The untagged VLAN ID
+  vlanId:
+  ownerRef:
+    name: sample-switch
+status:
+  // ConfigurationStatus indicates whether the switchPort is
+  // configured. When the network controller modifies the vlan
+  // information, it will set this value to notready.
+  // After the switch controller configuration is complete,
+  // this value will be set to ready.
+  ConfigurationStatus: ready
+```
+
+#### Switch Controller
+
+- Watch
+  - Switch CR
+  - SwitchPort CR
+- Configuration Operations
+  - When a new Switch CR is created, Switch Controller obtains all the ports of
+    the switch and their corresponding VLAN information, and creates a
+    SwitchPort CR for each port.
+  - When a SwitchPort CR's network configuration changes, Switch Controller
+    configures the corresponding port of the switch and sets
+    `.status.configurationStatus` to `ready` if it observed that the
+    configuration in the switch matches the spec of the SwitchPort CR.
+
+### Implementation Details/Notes/Constraints
+
+TBD
+
+### Risks and Mitigations
+
+TBD
+
+### Work Items
+- Define the Switch, SwitchPort, Network and Interface CRD
+- Implementation of Switch, Network and Interface Controller
+- Change the API of Metal3Cluster and BareMetalHost
+- Unit tests
+- E2e tests in metal3-dev-env
```

Comment:
This proposal does not say anything about a test setup. Metal3-dev-env switches are for now not configurable. it means we would need to switch to OVS probably (the setup with libvirt must be checked), and the maybe run an SDN controller on top, to not directly interact with OVS ? However, metal3-dev-env should probably still support deployments without the network APIs part, i.e. as it is now, being backwards compatible. This kind of changes should be addressed in the proposal too.